test/test_transformers.py

```diff
@@ -849,6 +849,42 @@ def test_encoder_is_causal(self):
 
         self.assertEqual(masked_output, is_causal_output)
 
+    @onlyCUDA
+    def test_math_backend_high_precision(self):
+        xq = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 5
+        xk = torch.rand([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16) * 5
+        xv = torch.randn([1, 6432, 2, 80], device="cuda", dtype=torch.bfloat16)
+        mask = None
+
+        def scaled_dot_product_attention(
+            xq: torch.Tensor, xk: torch.Tensor, xv: torch.Tensor, mask: Optional[torch.Tensor], backend: SDPBackend
+        ) -> torch.Tensor:
+            n_rep = 1
+            xq, xk, xv = [tensor.transpose(1, 2) for tensor in (xq, xk, xv)]
+            xk = xk.repeat_interleave(n_rep, dim=1)
+            xv = xv.repeat_interleave(n_rep, dim=1)
+
+            with sdpa_kernel(backends=[backend]):
+                attn_output = F.scaled_dot_product_attention(
+                    xq, xk, xv, attn_mask=mask, dropout_p=0.0
+                )
+            return attn_output.transpose(1, 2)
+
+        torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(True)
+        sdp_math_low_prec_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.MATH)
+        torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(False)
+        sdp_math_high_prec_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.MATH)
+        sdp_flash_out = scaled_dot_product_attention(xq, xk, xv, mask, SDPBackend.FLASH_ATTENTION)
+
+        torch.testing.assert_close(sdp_math_high_prec_out, sdp_flash_out, atol=1e-2, rtol=1e-2)
+
+        try:
+            torch.testing.assert_close(sdp_math_low_prec_out, sdp_flash_out, atol=1e-2, rtol=1e-2)
```

Comment:
as is, both of these assertions could pass and we wouldn't know whether the flag controls precision or not. see my prior suggestion for what I'm looking for from this test

Reply:
@jbschlosser for the 2nd assert on sdp_math_low_prec_out, we actually want to assert failure with try / except.